Port Qunit tests to cdp

larskarlitski opened the pull request:
Port QUnit tests to chrome debug protocol, reusing @martinpitt's work for the integration tests.

Marked as WIP, because:
- [x] add chromium to unit-test container for semaphore (#8349)
- [x] figure out how to subscribe to resource loading errors over CDP (see TODO in tap-cdp.py)
- [ ] remove workarounds we have for phantom behaving weirdly (if possible)
- [x] land #8350

followup cleanups:
- [ ] clean up injected scripts: move them from cdp.py to testlib.py and drop bool parameter

martinpitt commented on Makefile.am:
```diff
@@ -254,11 +254,7 @@ TEST_EXTENSIONS = .html .sh
-else
-HTML_LOG_DRIVER_FLAGS = --missing=phantomjs
-endif
+HTML_LOG_COMPILER = $(top_srcdir)/test/common/tap-cdp --strip=$(top_srcdir)/ -- $(HTML_TEST_SERVER)
```
You still need the `if`fery, as our test VM image build environments (mock, pbuilder) don't have chromium installed (nor phantomjs).

martinpitt commented:
@larskarlitski: Nice, thanks for working on this! I didn't review yet, but wrt. your questions about resource errors: They appear in [Log.entryAdded](https://chromedevtools.github.io/devtools-protocol/1-3/Log/#event-entryAdded), like [here](http://209.132.184.41/logs/pull-8069-20171208-065252-ec5d7ceb-verify-rhel-7/log.html#80):
```
CDP: {"source":"network","level":"error","text":"Failed to load resource: the server responded with a status of 401 (Authentication failed)","timestamp":1512717056493.03,"url":"http://127.0.0.2:9191/cockpit/login","networkRequestId":"26842.8"}
```
(of course this instance is an "expected" one). This is printed by the handler in test/common/cdp-driver.js.

Another promising API is [Network.loadingFailed](https://chromedevtools.github.io/devtools-protocol/1-2/Network/#event-loadingFailed), but I didn't experiment with that yet.

martinpitt commented on configure.ac:
```diff
@@ -269,12 +269,6 @@ else
-
-AC_PATH_PROG([PHANTOMJS], [phantomjs], [], [$srcdir/node_modules/.bin$PATH_SEPARATOR$PATH])
-AM_CONDITIONAL([WITH_PHANTOMJS], [test -n "$PHANTOMJS"])
-AC_SUBST([PHANTOMJS])
```
This needs to be replaced with logic to search for chromium, similar to [browser_path() in testlib.py](https://github.com/cockpit-project/cockpit/blob/master/test/common/testlib.py#L1265)

larskarlitski replied:
Indeed, thanks. Fixed this first to make the environments without chromium installed work.

martinpitt commented:
Thanks @larskarlitski ! I added a TODO item to add chromium to the unit-test container for semaphore.

martinpitt commented:
Why do you need to split out the `CDP` class, OOI? Can't you just import the whole testlib from the tap driver?

larskarlitski commented:
> Why do you need to split out the CDP class, OOI? Can't you just import the whole testlib from the tap driver?

To be honest, I can't think of the original reason anymore. I still think it's a bit cleaner, because the unit testing has no other common code with the integration tests. If we were to split out the integration tests at some point, this would make it easier. I don't feel strongly about it though and don't mind changing it if you prefer.

martinpitt commented:
@larskarlitski : No, it's fine for stylistic reasons, too. I just wondered if I was missing something.

martinpitt commented:
@larskarlitski : In PR #8349 I added chromium to the unit-tests container for semaphore, so you can start running tests there.

larskarlitski commented:
Fixed the `jsquote` issue (I moved it into cdp.py, but testlib.py uses it in other places as well) and separated the refactor into its own pull request at #8350.

martinpitt commented on bots/images/scripts/fedora-26.setup:
```diff
@@ -103,7 +103,6 @@ dnf $DNF_OPTS -y upgrade
 dnf $DNF_OPTS -y install $TEST_PACKAGES $COCKPIT_DEPS $IPA_CLIENT_PACKAGES $ABRT_PACKAGES $AVOCADO_TEST_DEPS
 
 # for avocado/fedora integration test
-npm -g install phantomjs-prebuilt
```
I'd suggest to leave the removal of phantomjs from the images to a follow-up PR, to avoid having to rebuild/land images in this PR.

larskarlitski replied:
Makes sense.

martinpitt commented on configure.ac:
```diff
@@ -269,11 +269,14 @@ else
-AC_PATH_PROG([PHANTOMJS], [phantomjs], [], [$srcdir/node_modules/.bin$PATH_SEPARATOR$PATH])
-AM_CONDITIONAL([WITH_PHANTOMJS], [test -n "$PHANTOMJS"])
-AC_SUBST([PHANTOMJS])
+AC_PATH_PROGS([CHROMIUM], [chromium-browsr chrome], [], [$srcdir/node_modules/chromium/lib/chromium/chrome-linux/chrome$PATH_SEPARATOR$PATH])
```
"browsr" typo

larskarlitski replied:
I think we should rename the package! ;)

martinpitt commented on test/common/tap-cdp:
```diff
@@ -0,0 +1,85 @@
+                          close_fds=True)
+
+cdp = cdp.CDP("C.utf8")
+cdp.invoke("Page.navigate", url='file:///home/lars/src/cockpit/dist/docker/test-docker.html')
```
I suppose this should be `opts.test` eventually? (I. e. this won't work on our CI like that)

larskarlitski replied:
Oooops, sorry. I used this for testing and forgot to remove it.

martinpitt commented on test/common/tap-cdp:
```diff
@@ -0,0 +1,85 @@
+    # TODO don't access private field
+    line = cdp._driver.stderr.readline()
+
+    _, message = driver_line_re.match(line).groups()
```
That's a bit ugly IMHO -- `.group(2)` instead?

larskarlitski replied:
Yeah that's much better, thanks. I also moved it into `cdp.py` to remove that one TODO.

martinpitt commented on test/common/tap-cdp:
```diff
@@ -0,0 +1,85 @@
+    if tap_line_re.match(message):
+        print(message)
+    else:
+        print(message, file=sys.stderr)
```
supposedly this should set `success = False` here

larskarlitski replied:
Yes, thanks.

martinpitt commented on test/common/cdp.py:
```diff
@@ -15,6 +15,8 @@
 
 TEST_DIR = os.path.normpath(os.path.dirname(os.path.realpath(os.path.join(__file__, ".."))))
 
+driver_line_re = re.compile(r'^> ([a-z]+): (.*)\n$')
```
shouldn't this be `console_line_re` rather?

martinpitt replied:
See below, this can just be dropped.

martinpitt commented on test/common/cdp.py:
```diff
@@ -228,3 +230,10 @@ def set_frame(self, frame):
+            return
+
+        line = self._driver.stderr.readline()
+        return driver_line_re.match(line).group(2)
```
This looks a bit fiddly. Also it cannot work as the `self._driver` process does not redirect its stderr. This is deliberate, as we want the user to see these messages from the driver.

If this is supposed to grab the first (or last) console message, then use `self.get_js_log()[0]` (or `[-1]` for the last one).

larskarlitski replied:
> This looks a bit fiddly. Also it cannot work as the self._driver process does not redirect its stderr. This is deliberate, as we want the user to see these messages from the driver.

It does as of #8350. We need this for the unit tests, because the tap tests output control messages such as `cockpittest-tap-done`, which we need to catch in `tap-cdp`.

I see your point for wanting to see these messages for the integration tests, though. We could either add an `stderr` parameter to the cdp class or find a different way of communicating status from the qunit tests.

martinpitt commented:
> It does as of #8350

Argh! I didn't see that, and the commit message didn't mention that either. Can you please revert this? What's wrong with grabbing the console log messages from `CDP.get_js_log()`?

larskarlitski commented:
Rebased. Now that `tap-cdp` cannot block on `cdp.stderr` anymore, we need to poll it via `get_js_log` (which I moved to the cdp class for this reason).

martinpitt commented on test/common/cdp-driver.js:
```diff
@@ -70,7 +71,10 @@ function setupLogging(client) {
     client.Log.enable();
-    client.Log.entryAdded(entry => process.stderr.write("CDP: " + JSON.stringify(entry["entry"]) + "\n"));
+    client.Log.entryAdded(entry => {
+        logEntries.push(entry["entry"]);
```
Is this related to the unit tests? I thought they use `console.log()`. The CDP `Log` is for messages that Chromium itself generates, not from the JS client script.

martinpitt replied:
Nevermind, supposedly this is for picking up Resource Errors.

martinpitt commented on test/common/cdp.py:
```diff
@@ -170,7 +170,7 @@ def start(self):
             # sandboxing does not work in Docker container
             self._browser = subprocess.Popen(
-                argv + ["--disable-gpu", "--no-sandbox", "--remote-debugging-port=%i" % cdp_port, "about:blank"],
+                argv + ["--disable-gpu", "--no-sandbox", "--disable-web-security", "--remote-debugging-port=%i" % cdp_port, "about:blank"],
```
This disables Same-Origin Policy, doesn't it? This will hide actual bugs, so if the unit tests need this for some reason, it should be a (non-default) option of the `CDP` class.

martinpitt commented:
I agree that the polling and multiplexing between Log and console are ugly. Some random ideas:

 * Re-revert commit 024818d2043 (`test: Revert redirecting the cdp driver's stderr`) and make it an option in the `CDP` constructor. Then integration tests work as before, and unit tests can set the option to read Console (`> ...`) and Log (`CDP: ...`) messages in order from stderr, without polling.

   Note that the `CDP:` prefix for Log isn't magic in any way. I didn't want to use the same prefix as for console (`>`) just to be able to tell them apart. I think this can be dealt with by the regexp matches (accept either prefix).

 * Outputting as JSON instead and always writing to stderr would be okay for me as well, but you then need a separate thread in `class CDP` to read and translate them. I don't like this much TBH, as it could cause messages to be printed out of order, which makes tests harder to debug.

 * Introduce a `CDP.wait_log()` method which doesn't immediately return with the current set of messages (like `get_*()`), but just sets up a promise that gets resolved by both `Log.entryAdded` and `Runtime.consoleAPICalled`. `CDP.command()` is already structured so that it blocks for the promise to get resolved. The resolving can then contain the actual message, so that on the `tap-cdp` side you would just need to call this in a simple loop.

larskarlitski commented:
Changed as discussed offline with @martinpitt: I've added a cdp.read_log() function which blocks until log messages arrive. This makes `tap-cdp` much easier to read.

I've also removed `--disable-web-security` that I had left in accidentally and made `tap-cdp` not fail anymore when one of the tests fails. This seems counter-intuitive for me, but it consistent with the previous behavior.

I've also added a fix for a hanging test in `src/base1/test-framed-cache.js`. @stefwalter, could you please have a look at that? I'm pretty sure that we can remove the whole `postMessage` thing and just call test.done() when we received all values in the consumer, but I'm not familiar with the semantics of the cache.  

Still WIP because some other tests are still hanging.

larskarlitski commented:
Update to print messages with `encode('utf-8')`.

Still not happy about the duplicated TAP output. Otherwise tests pass locally now, so I'm removing the WIP.

larskarlitski commented:
Update to not inject `sizzle.js` and `test_functions.js` when running the unit tests, because the paths are wrong during `make distcheck`. Thanks for helping figure this out, @martinpitt!

martinpitt commented on test/common/cdp.py:
```diff
@@ -226,3 +228,25 @@ def set_frame(self, frame):
+
+    def read_log(self):
+        """Returns an iterator that produces log messages one by one. Blocks if
+        there are no new messages right now."""
```
Short description of a docstring needs to be a single line. I. e. put a paragraph between the sentences.

larskarlitski replied:
+1

martinpitt commented on test/common/cdp-driver.js:
```diff
@@ -59,20 +59,56 @@ function success(result) {
+ * Only one such promise can be active at a given time. Once the promise is
+ * resolved, this function can be called again to wait for further messages.
+ */
+function waitLog() {
```
Maybe add an assertion that `logPromiseResolver` is undefined? (For robustness and easier debugging)

larskarlitski replied:
Yep!

martinpitt commented on test/common/cdp.py:
```diff
@@ -226,3 +228,25 @@ def set_frame(self, frame):
+        while True:
+            messages = self.command("waitLog()")
+            for m in messages:
+                yield m
```
It feels a bit odd that this is an infinite loop. But the caller exits it at an appropriate time, and I don't have a better idea about it either, so let's go with this for now.

larskarlitski replied:
Agreed, I couldn't think of a better way of doing this...

martinpitt commented on test/common/tap-cdp:
```diff
@@ -0,0 +1,86 @@
+import time
+
+tap_line_re = re.compile(r'^(ok [0-9]+|not ok [0-9]+|bail out!|[0-9]+\.\.[0-9]+|# )', re.IGNORECASE)
+console_line_re = re.compile(r'^([a-z]+): (.*)$')
```
This is unused.

larskarlitski replied:
Thanks!

martinpitt commented on test/common/tap-cdp:
```diff
@@ -0,0 +1,86 @@
+
+    # fail on browser level errors
+    if t == 'cdp':
+        if message['level'] != "error" or (ignore_resource_errors and "Failed to load resource" in message["text"]):
```
It took me a while to mentally invert the condition to verify it, and I'm still not sure about it. Maybe actually invert it and do the abort in the `then` branch, to make this a bit easier to read?

I was thinking it would still be useful to print the message to stderr so that it appears in the log. However, think they actually do already, as we don't redirect cdp-driver.js's  stderr. But doesn't that mean that `make check` would *also* show all the `tap-*` console messages, as they also get written to stderr? It doesn't hurt much, but it might look a bit untidy. Indeed I see a lot of lines like
```
PASS: dist/kubernetes/scripts/test-volumes.html 69 - nfs volume validate: no errors when valid
> log: # module: 
> log: # test: ip_prefix_from_text
> log: ok 1
> log: ok 2
```
in `make check` output.

That said, fine for me to clean this up in a follow-up.

larskarlitski replied:
Yeah that's what I meant with "duplicate output" in an earlier comment. Would be cool if we could address this separately.

larskarlitski replied:
Oh, and I've cleaned up the condition.

martinpitt commented on test/common/tap-cdp:
```diff
@@ -0,0 +1,86 @@
+    if t == 'cdp':
+        if message['level'] != "error" or (ignore_resource_errors and "Failed to load resource" in message["text"]):
+            continue
+        sys.exit(1)
```
Using `sys.exit` to  fail the test is unclean, as it will skip the `server.terminate` and `cdp.kill`. Maybe go back to what you had before with a `success` flag and use `break` instead?

larskarlitski replied:
Indeed! Thanks

larskarlitski commented:
> Also, please add a Closes # to the main commit and capitalize the commit messages, so that this becomes button-mergeable.

Done.

Thanks for the review. I think I've addressed all your points in the latest push.

martinpitt commented on test/common/tap-cdp:
```diff
@@ -0,0 +1,104 @@
+            if ignore_resource_errors and "Failed to load resource" in message["text"]:
+                continue
+            success = False
+            break
```
Much simpler to understand now, even though it's a bit more verbose. Thanks!
